File: android/app/src/main/java/com/openburnbar/services/media/AIInboxNotificationRouting.kt
package com.openburnbar.services.media

import android.app.Notification
import android.app.PendingIntent
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.openburnbar.BurnBarDeepLink
import com.openburnbar.MainActivity

/**
 * AI Inbox P1 alerts, delivered as `type=ai_inbox_item` FCM data messages by
 * `functions/src/aiInboxNotifications.ts`.
 *
 * ## What the push does and does not contain
 *
 * The mirrored inbox document is sealed with the Cloud Vault, so the Cloud
 * Function that fires this push cannot read the item's title or summary and
 * therefore cannot put them in the payload. What arrives is an opaque `item_id`
 * plus two bounded enums (`kind`, `priority`). The notification body is
 * deliberately generic; tapping it deep-links into the app, which fetches and
 * decrypts the real content locally.
 *
 * That means this file must never try to render item text from the payload —
 * there is none to render, and treating a future server-supplied string as
 * content would quietly turn the alert into a plaintext leak channel.
 */
internal data class AIInboxNotificationRouting(
    val itemId: String,
    val kind: String,
    val title: String,
    val body: String,
    val deepLink: String,
    val eventId: String? = null,
    val uid: String? = null,
    val expiresAtMillis: String? = null,
    val type: String? = null,
)

/** `BurnBarInboxItemKind` raw values, mirrored from the Kernel contract. */
private val INBOX_KIND_TITLES =
    mapOf(
        "ci_waste" to "Wasted CI",
        "promised_not_landed" to "Promised, not landed",
        "uncommitted_work" to "Uncommitted work",
        "cost_anomaly" to "Cost anomaly",
        "stuck_pr" to "Stuck pull request",
        "index_health" to "Index health",
        "brief" to "Brief",
        "budget" to "Budget",
        "system" to "OpenBurnBar",
    )

private const val FALLBACK_KIND = "system"
private const val GENERIC_BODY = "Open OpenBurnBar to see what needs you."
private const val MAX_ITEM_ID_LENGTH = 160

// Control/bidi characters in an id would let a hostile payload spoof the
// rendered notification; the same guard the Mercury routing ids use.
private val CONTROL_OR_BIDI = Regex("[\\u0000-\\u001f\\u202a-\\u202e]")

/**
 * Resolves an `ai_inbox_item` payload into what the notification should show.
 *
 * Returns `null` when the payload cannot address a real item, because a
 * notification the user cannot act on is worse than none: the tap would land on
 * an empty surface with no way back to what fired it.
 *
 * Pure and internal so the dispatch contract is unit-testable without an
 * emulator — the push shape is a cross-repo contract with the Cloud Function,
 * and a silent drift in it produces no crash, just notifications that stop
 * arriving.
 */
internal fun aiInboxNotificationRouting(data: Map<String, String>): AIInboxNotificationRouting? {
    val itemId = normalizedItemId(data["item_id"]) ?: return null
    val rawKind = data["kind"]?.trim().orEmpty()
    val kind = if (INBOX_KIND_TITLES.containsKey(rawKind)) rawKind else FALLBACK_KIND
    return AIInboxNotificationRouting(
        itemId = itemId,
        kind = kind,
        // Server-supplied `title` is only ever a kind label, so prefer the local
        // table: it keeps the string translatable and makes it impossible for a
        // payload field to become a content channel.
        title = INBOX_KIND_TITLES.getValue(kind),
        body = GENERIC_BODY,
        deepLink = BurnBarDeepLink.inboxURL(itemId),
        eventId = data["event_id"]?.trim()?.takeIf { it.isNotBlank() },
        uid = data["uid"]?.trim()?.takeIf { it.isNotBlank() },
        expiresAtMillis = data["expires_at_millis"]?.trim()?.takeIf { it.isNotBlank() },
        type = data["type"]?.trim()?.takeIf { it.isNotBlank() },
    )
}

private fun normalizedItemId(raw: String?): String? {
    val trimmed = raw?.trim()?.takeIf { it.isNotBlank() } ?: return null
    if (trimmed.length > MAX_ITEM_ID_LENGTH) return null
    if (CONTROL_OR_BIDI.containsMatchIn(trimmed)) return null
    return trimmed
}

internal fun MercuryFcmService.buildAIInboxNotification(routing: AIInboxNotificationRouting): Notification {
    ensureInboxChannel()
    val openIntent =
        Intent(this, MainActivity::class.java).apply {
            action = Intent.ACTION_VIEW
            data = Uri.parse(routing.deepLink)
            routing.eventId?.let { putExtra(MainActivity.EXTRA_EVENT_ID, it) }
            routing.uid?.let { putExtra(MainActivity.EXTRA_UID, it) }
            routing.expiresAtMillis?.let { putExtra(MainActivity.EXTRA_EXPIRES_AT_MILLIS, it) }
            routing.type?.let { putExtra(MainActivity.EXTRA_PUSH_TYPE, it) }
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        }
    // Pinned on its own line, outside the `apply` block, matching
    // MercuryFcmServiceSupport: the explicit-PendingIntent gate verifies the
    // package pin as a directly greppable statement per intent variable.
    openIntent.setPackage(packageName)
    val openPending =
        PendingIntent.getActivity(
            this,
            routing.itemId.hashCode(),
            openIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )

    return NotificationCompat.Builder(this, MercuryFcmService.INBOX_CHANNEL_ID)
        .setSmallIcon(android.R.drawable.stat_sys_warning)
        .setContentTitle(routing.title)
        .setContentText(routing.body)
        .setContentIntent(openPending)
        .setAutoCancel(true)
        .setCategory(NotificationCompat.CATEGORY_REMINDER)
        .setPriority(NotificationCompat.PRIORITY_HIGH)
        // No public version: even the kind label describes the user's own work,
        // so a locked screen shows the channel name and nothing else.
        .setVisibility(NotificationCompat.VISIBILITY_PRIVATE)
        .setLocalOnly(true)
        .build()
}

internal fun MercuryFcmService.postAIInboxNotification(data: Map<String, String>) {
    val routing = aiInboxNotificationRouting(data) ?: return
    try {
        NotificationManagerCompat.from(this).notify(routing.itemId.hashCode(), buildAIInboxNotification(routing))
    } catch (error: SecurityException) {
        // Missing POST_NOTIFICATIONS — record the real (revoked) permission
        // state so the cloud fan-out stops targeting this device rather than
        // counting silently-dropped pushes as delivered.
        Log.w("BurnBar", "ai_inbox_notification_post_denied kind=${routing.kind} reason=${error.message}")
        AgentReplyNotificationState.recordPermissionResult(applicationContext, granted = false)
    }
}

File: OpenBurnBarMobile/Services/AIInboxNotificationRouting.swift
import Foundation
import OpenBurnBarKernel

/// The AI Inbox half of the push-notification plane.
///
/// ## What a P1 push contains
///
/// The mirrored inbox document at `users/{uid}/ai_inbox_items/{itemId}` is
/// sealed with the Cloud Vault, so the Cloud Function that fires the push
/// (`functions/src/aiInboxNotifications.ts`) cannot read the item's title or
/// summary and therefore cannot put them in the payload. What arrives is an
/// opaque `item_id` plus two bounded enums — `kind` and `priority`.
///
/// This type deliberately exposes no `preview`/`summary` field: the alert body
/// is a fixed local string, and the real content is decrypted from Firestore
/// once the tap lands. A future server-supplied string rendered as content would
/// quietly convert the alert into a plaintext channel.
struct AIInboxNotificationPayload: Equatable, Sendable {
    /// The mirror document id. Opaque to the notification layer.
    let itemID: String
    /// A `BurnBarInboxItemKind` raw value, already narrowed to a known case.
    let kind: BurnBarInboxItemKind
    let priority: BurnBarInboxPriority
    /// The durable `agent_notification_events` id, for delivery correlation.
    let eventID: String
    let uid: String?
    let expiresAtMs: Int64?

    /// Kind label shown as the notification title. Local, never server text.
    var title: String { Self.title(forKind: kind) }

    /// Body copy. Fixed: there is nothing readable to summarize.
    var body: String { "Open OpenBurnBar to see what needs you." }

    /// Main-actor because `AIInboxDeepLink` owns a type-level stash. The payload
    /// itself stays non-isolated so it can be decoded inside the `nonisolated`
    /// `UNUserNotificationCenterDelegate` hops.
    @MainActor
    var deepLink: URL? { AIInboxDeepLink.url(itemID: itemID) }

    init?(userInfo: [AnyHashable: Any]) {
        guard Self.string(userInfo["type"]) == AIInboxDeepLink.pushType else { return nil }
        guard let raw = Self.string(userInfo["item_id"]),
              let itemID = AIInboxDeepLink.sanitizedItemID(raw) else {
            return nil
        }
        self.itemID = itemID
        // An unrecognized kind means a newer Mac published a case this build
        // does not know. Degrading to `.system` keeps the alert deliverable
        // instead of dropping a real P1 on a schema bump.
        if let rawKind = Self.string(userInfo["kind"]), let known = BurnBarInboxItemKind(rawValue: rawKind) {
            kind = known
        } else {
            kind = .system
        }
        priority = BurnBarInboxPriority(clamping: Int(Self.string(userInfo["priority"]) ?? "") ?? 1)
        eventID = Self.string(userInfo["event_id"]) ?? itemID
        uid = Self.string(userInfo["uid"]) ?? Self.string(userInfo["account_uid"])
        expiresAtMs = Self.string(userInfo["expires_at_millis"]).flatMap { Int64($0) }
    }

    static func title(forKind kind: BurnBarInboxItemKind) -> String {
        switch kind {
        case .ciWaste: return "Wasted CI"
        case .promisedNotLanded: return "Promised, not landed"
        case .uncommittedWork: return "Uncommitted work"
        case .costAnomaly: return "Cost anomaly"
        case .stuckPR: return "Stuck pull request"
        case .indexHealth: return "Index health"
        case .brief: return "Brief"
        case .budget: return "Budget"
        case .system: return "OpenBurnBar"
        }
    }

    private static func string(_ value: Any?) -> String? {
        switch value {
        case let string as String:
            let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? nil : trimmed
        case let number as NSNumber:
            return number.stringValue
        default:
            return nil
        }
    }
}

/// `burnbar://inbox` routing, shared by the push tap and the URL handler.
///
/// Follows `HermesGatewayPairingDeepLink`: the tap and the app launch race each
/// other, so the request is stashed and posted, and whichever surface is alive
/// claims it. `consumePendingItemID` clears as it reads so a re-render cannot
/// re-navigate away from wherever the user has since gone.
///
/// `@MainActor` for that type-level stash, matching the precedent. Every caller
/// (URL handler, banner tap, notification-response hop) is already isolated.
@MainActor
enum AIInboxDeepLink {
    // Constants and pure parsing stay `nonisolated`: `AIInboxNotificationPayload
    // .init(userInfo:)` decodes inside a `nonisolated`
    // `UNUserNotificationCenterDelegate` hop, so anything it touches must be
    // reachable without hopping to the main actor first.
    nonisolated static let notificationName = Notification.Name("ShowAIInboxItem")
    nonisolated static let itemIDUserInfoKey = "itemID"
    nonisolated static let host = "inbox"
    nonisolated static let pushType = "ai_inbox_item"

    /// Bounds an id before it becomes a URL path or a lookup key.
    nonisolated private static let maxItemIDLength = 160

    private static var pendingItemID: String?

    static func open(itemID: String?) {
        let sanitized = itemID.flatMap(sanitizedItemID)
        pendingItemID = sanitized
        var userInfo: [AnyHashable: Any] = [:]
        if let sanitized { userInfo[itemIDUserInfoKey] = sanitized }
        NotificationCenter.default.post(name: notificationName, object: nil, userInfo: userInfo)
    }

    nonisolated static func itemID(from notification: Notification) -> String? {
        (notification.userInfo?[itemIDUserInfoKey] as? String).flatMap(sanitizedItemID)
    }

    static func consumePendingItemID() -> String? {
        let itemID = pendingItemID
        pendingItemID = nil
        return itemID
    }

    /// `burnbar://inbox/<itemId>`, or the bare list link when there is no item.
    nonisolated static func url(itemID: String?) -> URL? {
        guard let sanitized = itemID.flatMap(sanitizedItemID) else {
            return URL(string: "burnbar://\(host)")
        }
        let encoded = sanitized.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? sanitized
        return URL(string: "burnbar://\(host)/\(encoded)")
    }

    /// Extracts the item id from a `burnbar://inbox[/<itemId>]` URL.
    nonisolated static func itemID(from url: URL) -> String? {
        guard url.scheme?.lowercased() == "burnbar", url.host?.lowercased() == host else { return nil }
        return url.pathComponents.dropFirst().first.flatMap(sanitizedItemID)
    }

    /// Rejects ids that are empty, oversized, or carry control/bidi characters —
    /// the last of which would let a hostile payload spoof rendered text.
    nonisolated static func sanitizedItemID(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.count <= maxItemIDLength else { return nil }
        guard trimmed.rangeOfCharacter(from: .controlCharacters) == nil else { return nil }
        guard trimmed.unicodeScalars.allSatisfy({ !(0x202A...0x202E).contains($0.value) }) else { return nil }
        return trimmed
    }

    /// Test seam: the type-level stash must not leak between test cases.
    static func resetPendingItemID() {
        pendingItemID = nil
    }
}
